### PR TITLE
[ready] Add multivariate log-gamma (mvlgamma)

### LINKED ISSUE #9378: [feature request] Multivariate Gamma function
Looks like there currently is no implementation of the (log-) multivariate Gamma function? That would be useful for implementing Wishart / inverse Wishart distributions.

The straightforward way to do this would be to reduce it to multiple `lgamma` evaluations using the identity 35.3.4 from https://dlmf.nist.gov/35.3. For a float `a` the `p`-dim multivariate Gamma function in python is just

```
def log_mv_gamma(p, a):
    C = p * (p - 1) / 4 * math.log(math.pi)
    return C + torch.lgamma(a - 0.5 * torch.arange(p, dtype=torch.float)).sum()
```

I'm not sure how numerically stable this simple approach is, but it doesn't look too bad at first glance.

### PR DESCRIPTION
**Pull request by vishwakftw**
1. Add tests in test_cuda, test_torch
2. Add doc strings

Closes https://github.com/pytorch/pytorch/issues/9378 .

cc: @alicanb

### CODE REVIEW

File: aten/src/ATen/native/UnaryOps.cpp [hunk 1]
```diff
@@ -32,6 +38,28 @@ Tensor& fill_(Tensor& self, const Tensor& value) {
   return self._fill_(value);
 }
 
+Tensor mvlgamma(const Tensor& self, int64_t p) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()),
+           "mvlgamma is not implemented for ", self.type());
+  AT_CHECK((self > 0.5 * (p - 1.)).all().toCByte(),
+           "Condition for computing multivariate log-gamma not met");
+  AT_CHECK(p >= 1, "p has to greater than or equal to 1");
```
Comment (ezyang) [inline aten/src/ATen/native/UnaryOps.cpp]: has to be

Comment (vishwakftw) [reply aten/src/ATen/native/UnaryOps.cpp]: Oops, sorry. Has been fixed the following commit. Thank you.

File: aten/src/ATen/native/UnaryOps.cpp [hunk 2]
```diff
@@ -32,6 +38,28 @@ Tensor& fill_(Tensor& self, const Tensor& value) {
...
 }
 
+Tensor mvlgamma(const Tensor& self, int64_t p) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()),
+           "mvlgamma is not implemented for ", self.type());
+  AT_CHECK((self > 0.5 * (p - 1.)).all().toCByte(),
+           "Condition for computing multivariate log-gamma not met");
+  AT_CHECK(p >= 1, "p has to be greater than or equal to 1");
+  Tensor args = native::arange(-p + 1, 1, self.options()).div_(2.);
+  args = args.add(self.unsqueeze(-1));
```
Comment (SsnL) [inline aten/src/ATen/native/UnaryOps.cpp]: `add_`  seems better?

File: aten/src/ATen/native/UnaryOps.cpp [hunk 3]
```diff
@@ -32,6 +38,28 @@ Tensor& fill_(Tensor& self, const Tensor& value) {
   return self._fill_(value);
 }
 
+Tensor mvlgamma(const Tensor& self, int64_t p) {
+  AT_CHECK(at::isFloatingType(self.type().scalarType()),
+           "mvlgamma is not implemented for ", self.type());
+  AT_CHECK((self > 0.5 * (p - 1.)).all().toCByte(),
+           "Condition for computing multivariate log-gamma not met");
+  AT_CHECK(p >= 1, "p has to be greater than or equal to 1");
+  Tensor args = native::arange(-p + 1, 1, self.options()).div_(2.);
```
Comment (SsnL) [inline aten/src/ATen/native/UnaryOps.cpp]: nit: use `step=0.5`

File: aten/src/ATen/native/UnaryOps.cpp (hunk 2 above)
Comment (vishwakftw) [reply aten/src/ATen/native/UnaryOps.cpp]: I recall facing a problem with `add_` (a broadcasting issue), which is why I resorted to this.

File: aten/src/ATen/native/UnaryOps.cpp (hunk 3 above)
Comment (vishwakftw) [reply aten/src/ATen/native/UnaryOps.cpp]: Oh yes, lesser ops. Thank you!!

File: aten/src/ATen/native/UnaryOps.cpp (hunk 2 above)
Comment (SsnL) [reply aten/src/ATen/native/UnaryOps.cpp]: I see. How about `add_(self.unsqueeze(-1).expand_as(arg))s`?

Comment (vishwakftw) [reply aten/src/ATen/native/UnaryOps.cpp]: That solution is perhaps incorrect because it would effectively make `self` the same size as `args`.

Comment (SsnL) [reply aten/src/ATen/native/UnaryOps.cpp]: ah i see. it's the `args` that should be broadcasted, not `self.

**Comment by vishwakftw** [comment]
Is this good to go? Sorry about the reminder.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**